### PR TITLE
First cut of adding support for types registry. 

### PR DESCRIPTION
**Pull request by vaikas**
- Create types/redis/v1 and types/replicatedservice/v1 directories and copy the example templates there
- Introduce registry.go and implement a github implementation of it that demonstrates how to use it
- First cut of implementing: #22 and #23.
- Does not support schema validation yet nor enforce that it exists (TODO)

### CODE REVIEW

File: client/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+const TypesDir string = "types"
+
+type Type struct {
+	Name string
+	Version string
+}
+
+// Registry abstracts type interactions.
+type Registry interface {
+	// List all the types in the given repository
```
Comment (jackgr) [inline client/registry/registry.go]: repository -> registry

File: client/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registry
+
+// Registry abstracts a types registry which holds types that can be
+// used in a Deployment Manager configurations. A registry root must have
```
Comment (jackgr) [inline client/registry/registry.go]: configurations -> configuration

File: client/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// a 'types' directory which contains all the available types. Each type
+// then contains version directory which in turn contains all the files
```
Comment (jackgr) [inline client/registry/registry.go]: directory -> directories

File: client/registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// necessary for that type.
```
Comment (jackgr) [inline client/registry/registry.go]: necessary for that version of the type.

File: client/registry/github_registry.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+package registry
+
+import (
+	"github.com/google/go-github/github"
+
+	"fmt"
+	"log"
+)
+
+type GithubRegistry struct {
```
Comment (jackgr) [inline client/registry/github_registry.go]: Public type needs comment.

File: client/registry/github_registry.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	owner      string
+	repository string
+	client     *github.Client
+}
+
+func NewGithubRegistry(owner string, repository string) *GithubRegistry {
```
Comment (jackgr) [inline client/registry/github_registry.go]: Public type needs comment.

File: client/registry/github_registry.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	return &GithubRegistry{
+		owner:      owner,
+		repository: repository,
+		client:     github.NewClient(nil),
+	}
+}
+
+func (g *GithubRegistry) List() ([]Type, error) {
```
Comment (jackgr) [inline client/registry/github_registry.go]: Public type needs comment.

File: client/registry/github_registry.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		for _, v := range versions {
+			retTypes = append(retTypes, Type{Name: t, Version: v})
+		}
+	}
+
+	return retTypes, nil
+}
+
+// Get the URL for a given type
+func (g *GithubRegistry) GetURL(t Type) (string, error) {
```
Comment (jackgr) [inline client/registry/github_registry.go]: Public method needs comment.

**Comment by jackgr** [comment]
Minor changes. Otherwise, LGTM.

File: client/client.go [hunk 9]
```diff
@@ -31,9 +32,10 @@ import (
 )
 
 var (
-	action  = flag.String("action", "deploy", "expand | deploy | list | get | delete | update | listtypes | listtypeinstances")
+	action  = flag.String("action", "deploy", "expand | deploy | list | get | delete | update | listtypes | listtypeinstances | types")
 	name    = flag.String("name", "", "Name of template or deployment")
 	service = flag.String("service", "http://localhost:8080", "URL for deployment manager")
+	type_registry = flag.String("type_registry", "kubernetes/deployment-manager", "Type registry [owner/repo], defaults to kubernetes/deployment-manager/")
```
Comment (jackgr) [inline client/client.go]: Nit: easier to just call it registry.

Comment (vaikas-google) [inline client/client.go]: registry is already used as a package name so that won't work, or my go-fu is weak...

**Comment by vaikas-google** [comment]
Thanks, PTAL. I believe I addressed all the comments, sans the type_registry / registry flag issue.

File: client/registry/github_registry.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import (
+	"github.com/google/go-github/github"
+
+	"fmt"
+	"log"
+)
+
+// GithubRegistry implements the Registry interface that talks to github.
+type GithubRegistry struct {
+	owner      string
```
Comment (jackgr) [inline client/registry/github_registry.go]: owner -> organization, since that's what Github calls it.

Comment (vaikas-google) [inline client/registry/github_registry.go]: Actually the github API calls it owner :)
https://github.com/google/go-github/blob/master/github/repos_contents.go#L130

**Pull request merged**